### ISSUE TITLE
Dataset Unzip Operation

### ISSUE THREAD

**Issue by eaplatanios** [tensorflow/tensorflow#12851]
@mrry This is a comment I originally posted to the "Redesigning input pipelines" issue, but I think it went unnoticed given the amount of comments on that thread. Given that issue is now closed (for good reason :)), I decided to post it as a separate issue.

I currently cannot see a way currently to "unzip" a dataset. Let's say we have a trainable model that has both a train/fit method and a infer/predict method. Let's call the type of the (potentially) nested structure of inputs to our model `I` and the type of training inputs, which are only needed when training (e.g., supervision labels), `TI`. In this case, we want the train method to accept datasets with elements of type `(I, TI)` (i.e., a tuple of `I` and `TI`) and the predict method to accept datasets with elements of type `I` or `(I, TI)` (in which case it would ignore the labels). We also want the model to only have one underlying graph, supporting all these types of input. The way I could see doing that was for the underlying model to construct two iterators (one with elements type `I` and one with type `TI`) and initialize them according to the provided datasets. However, if somebody provides a dataset with elements of type `(I, TI)` to the train method, there is no way to unzip this dataset and initialize both iterators. One has to use `Dataset.map` twice, which is not efficient (I think but please correct me if I'm wrong) and which may also not pull matching elements from the datasets (if each pull advances the current index in the original first dataset -- I'm not sure if that happens).

**Comment by mrry** [comment]
Thanks for reposting this @eaplatanios! I'd like to understand the use case more, so we can focus on the appropriate efficient solution. Just to check my understanding, are you proposing a transformation with the following type? (Please forgive the abuse of Haskell notation below :)....)

```haskell
unzip :: Dataset (I, TI) -> (Dataset I, Dataset TI)
```

If that's the case, you could do that by mapping twice over the the input, with the first map selecting the first component of the tuple, and the second map selecting the second component of the tuple. This would work—`Dataset` objects are stateless, and the state is in the `Iterator`—but you're right that it would be inefficient because (at least currently) the upstream work of producing the input (reading from files, parsing, etc.) would be duplicated.

How could we avoid that duplication? Today, the easiest way to do that would be by adding a `Dataset.cache()` before the `unzip`, so that the elements are computed once and subsequently read from the cache (either in memory or on disk). However, this only works with finite `Dataset` objects, which is not as featureful as a true `unzip`! Plus there's not (necessarily) any need to cache the entire dataset, depending on how you are planning to consume the iterators.

I'm a little unclear about why you'd want *two* iterators as input to the graph in your example. I'd presume that training and test data come from two different sources, so it wouldn't be a problem to create different iterators for those cases, and these could have the appropriate type `(I, TI)` or `(I)` as necessary. However, if you have a training graph and an input dataset of type `(I, TI)`, why would you prefer to unzip the datasets and make two iterators, rather than making one iterator and destructuring the elements it yields? The only reason I can think of—but I haven't thought too hard about this—is that you'd like to advance the iterators at different rates, but this creates a buffering issue: how many elements of type `TI` should be buffered while we're consuming the iterator of type `I` (or vice versa) to avoid recomputation?

However, just because I can't think of an application doesn't mean that there isn't one! I'd be interested to learn more about your feature request....

**Comment by eaplatanios** [comment]
@mrry Thanks for the fast and elaborate response! :)

First of all, let me provide a simplified example of the use case I have in mind. I want to create a functional API for constructing and training networks. I'm thinking of having a notion of `Layer[T, R]`, which maps from input type `T` to output type `R`. A layer is a function that given a `T` will construct relevant ops and return an `R`. These types could be tuples of op outputs, or other arbitrary structures for that matter. Layers can be combined in various ways (e.g., composed or mapped) to create new layers. A model is then defined using:
- Input layer that returns type `I`
- Inference layer that goes from `I` to `P` (think of this as the main model -- the prediction function -- e.g., a multi-layer perceptron)
- Train input layer that returns type `IT` (optional -- e.g., labels used when training a classifier)
- Train processing layer that goes from `IT` to `PT` (optional -- e.g., one-hot encoding of labels)
- Loss layer that goes from `(P, PT)` to a scalar tensor (optional -- e.g., cross-entropy)
- ...
The optional components are only used when training so if you create a model that requires no training, you can skip them. The model defines `train` and `predict` methods that accept data (train accepts a tuple of `(I, IT)`, but predict can accept either such a tuple, or simply data of type `I`).

Layers do not need to construct any ops before being "applied" to some input. Before creating a model, no interaction with TensorFlow is necessary and no ops are created. The TF graph is created, populated, and managed by the model. So, while initializing, the model will create a graph and use it for all its operations. I want the model to use the same graph for both training and prediction and I want to avoid adding redundant ops to that graph as much as possible. I think it's more elegant to limit graph management to the model so the user never has to deal with that (I've noticed it often confuses many people using TF).

**This point is important to our discussion:** Since I want to have a fixed graph, I need the inputs of that graph to be fixed at construction time and I want to use the dataset API iterators for that to allow flexibility with respect to how users provide data for training and/or prediction. Given this constraint, at graph construction time I need to either use two iterators (one for type `I` and one for `IT`), or an iterator over tuples `(I, IT)`. _If I use an iterator over tuples (which makes sense given that we want to advance the iterators at the same time), then how do I initialize that iterator with a dataset of type `I` for the model predict method?_

I hope that this simplified description of what I'm hoping to achieve makes sense.

With respect to your points, I have a couple of comments/questions:
- What you mention about the unzip operation and the semantics of traversing the unzipped iterators makes sense and I hadn't really thought about it. I always had the synchronized advancement of both in my mind. In this case, not unzipping them also makes sense as it incorporates that constraint in the structure.
- In general, I'm a bit unclear as to when the iterators are advanced. The simple use cases makes sense, but what about the following example: we have a model that has multiple entry points for a single input (i.e., multiple ops consuming the same input). So, during a single `Session.run` call, multiple ops will consume the same input provided by a single iterator. Will the iterator only advance once per `Session.run` call, or will it advance once for every single op that consumes its output? If the latter happens, then how can we enforce the once per `Session.run` call behavior?

Thank you for your response and I'm sorry for the super long post. I just want to try and make the setting clear so the use case may make sense. I do realize now though that unzipping may not be the right approach and I might be missing some simple solution to my example. :)

**Comment by eaplatanios** [comment]
I guess another not-so-pretty solution would be to have an iterator over tuples `(I, IT)` in my model, always, and then, if a user provides a dataset of `I`, I map that to one of `(I, TI)`, with dummy values for the `TI` elements. Although, I feel that this would be inefficient and also a bit ugly.

**Comment by svetlov** [comment]
@mrry  i also need `unzip`, cause of some transormations which expect dataset, containing only single component (i.e. tf.contrib.data.dense_to_sparse_batch).

**Comment by gibiansky** [comment]
Confirming this; `unzip` would be very useful.

In my application, I have to read data from JSON files and I extract two (or more) pieces of information for them, so I have a Dataset which yields (tf.string, tf.int32, tf.int32), for example. I then have some tensorflow code that transforms the tf.string into a tf.float32 (loading files, doing some compute), but leaves the two tf.int32's alone. Right now, I have to write a function like this:

```python
def transformer(a, b, c):
    a_new = fix_my_value(a)
    return a_new, b, c
```
But I would rather be able to unzip a dataset, apply `fix_my_value` directly, and then zip it back up.

**Comment by KingRando95** [comment]
Confirm

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by soloman817** [comment]
what is the status of this issue? I also need an unzip operation!

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by mrry** [comment]
Hi folks, I haven't had the time to study this problem, but it looks like several people would find it useful, so I'm going to throw it open to contributions.

**Comment by jhultman** [comment]
Manually unzipping doesn't seem to respect the order of the tuple elements when shuffling the zipped dataset. Take a look at my minimum working example below:

```
import tensorflow as tf

ds = tf.data.Dataset.range(100)
ds = tf.data.Dataset.zip((ds, ds))
ds = ds.shuffle(100)

x0 = ds.map(lambda *x: x[0])
x1 = ds.map(lambda *x: x[1])

ds = tf.data.Dataset.zip((x0, x1))

it = ds.make_one_shot_iterator()
tup = it.get_next()

with tf.Session() as sess:
    t1, t2 = sess.run(tup)
    print((t1, t2))
    
>> (8, 94)
```

I would have expected t1 == t2, since the shuffle was only performed on the zipped dataset. 

My use case is that I have input filenames and label filenames, and I want to shuffle the zipped dataset while maintaining the pairing between input and label. Later I want to unzip since I have a complicated input pipeline which processes inputs and labels separately.

Anyone have a suggestion for how to do this sort of shuffle? Seems like it should be an extremely common use case.

**Comment by matthai** [comment]
I agree, I find the behavior of `zip(ds, ds)` above unintuitive. The `dataset` framework seems to behavior less like a dataflow graph with mutable state and more like a pure function composition framework unlike the [core tensorflow framework](http://delivery.acm.org/10.1145/3090000/3088527/pldiws17mapl-maplmainid2.pdf?ip=131.107.174.21&id=3088527&acc=OA&key=4D4702B0C3E38B35%2E4D4702B0C3E38B35%2E4D4702B0C3E38B35%2E5945DC2EABF3343C&__acm__=1539194417_890fe79cd2079ceacf4d1229c27b4fee  ). Note that @mrry is a co-author on the above paper.

**Comment by BenjBarral** [comment]
> Manually unzipping doesn't seem to respect the order of the tuple elements when shuffling the zipped dataset. Take a look at my minimum working example below:
> 
> ```
> import tensorflow as tf
> 
> ds = tf.data.Dataset.range(100)
> ds = tf.data.Dataset.zip((ds, ds))
> ds = ds.shuffle(100)
> 
> x0 = ds.map(lambda *x: x[0])
> x1 = ds.map(lambda *x: x[1])
> 
> ds = tf.data.Dataset.zip((x0, x1))
> 
> it = ds.make_one_shot_iterator()
> tup = it.get_next()
> 
> with tf.Session() as sess:
>     t0, t1 = sess.run(tup)
>     print((t0, t1))
>     
> >> (8, 94)
> ```
> 
> I would have expected t0 == t1, since the shuffle was only performed on the zipped dataset.
> 
> My use case is that I have input filenames and label filenames, and I want to zip them, shuffle, and unzip while maintaining the pairing between input and label. I want to unzip since I have a complicated input pipeline which processes inputs and labels separately. Anyone have a suggestion for how to do this sort of shuffle? Seems like it should be an extremely common use case.
> 
> Edit1: I should probably just make a new issue, but I'm a bit astonished by the following simple example as well. Is this really the expected behavior? This seems a bit unintuitive.
> 
> Edit2: Passing an explicit nonzero seed to the shuffle solves the problem. But that defeats the entire purpose of my shuffle, which is to randomize the order every epoch. I suppose using the epoch itself as the seed could work, but seems like a very annoying work-around for such a simple problem.
> 
> ```
> import tensorflow as tf
> 
> ds = tf.data.Dataset.range(100).shuffle(100)
> ds = tf.data.Dataset.zip((ds, ds))
> tup = ds.make_one_shot_iterator().get_next()
> 
> with tf.Session() as sess:
>     print(sess.run(tup))
> 
> >> (42, 28)
> ```

Hi,
Have you found a way to shuffle a zipped dataset while still preserving the pairing ?
Thanks in advance

**Comment by essoca** [comment]
Hi guys, is there any progress on this issue? I also need an unzip operation!

**Comment by Santosh-Gupta** [comment]
Any update?

**Comment by claverru** [comment]
I just lost a day of work because of this :O

**Comment by jpauwels** [comment]
I stumbled on this discussion while searching for a built-in unzip operator, so I wanted to share my solution (actually two variants, can't decide which one is bettter). It works for shuffled datasets too, since it only iterates through the original dataset once.

You can use the Python unzip function as follows, which even works with batched datasets:
```feature_ds, label_ds = [tf.data.Dataset.from_tensor_slices(tf.concat(x, axis=0)) for x in zip(*ds)]```

Alternatively, for non-batched datasets you don't need the `tf.concat` (or you can `unbatch()` your `Dataset` first as below):
```feature_ds, label_ds = [tf.data.Dataset.from_tensor_slices(list(x)) for x in zip(*ds.unbatch())]```

The only drawback I can see is that both options need to load the entire `Dataset` into memory, so this won't work for very large datasets. That's why I still would like to see a native operator, to handle distributed `Datasets`.

My usecase is to split features from labels to create a confusion matrix, which I'd say is pretty common. Or am I missing a simpler solution for this?

**Comment by applenob** [comment]
We can use same seed to control the order of feature and label after shuffle. In this case we don't have to zip two datasets.

```
feature_ds.shuffle(size, seed=0).repeat()
label_ds.shuffle(size, seed=0).repeat()
```

**Comment by itcarroll** [comment]
A tf.data.Dataset unzip method could solve the problem of train-test splits on tf.data.Dataset.from_generator pipelines. Every other way I've found requires, impractically, iterating the dataset multiple times.

**Comment by sushreebarsa** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow version with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the release notes to stay up to date with the latest developments which are happening in the Tensorflow space.
Thank you!
